server/src/main/java/org/elasticsearch/index/fieldvisitor/IgnoredSourceFieldLoader.java

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+package org.elasticsearch.index.fieldvisitor;
+
+import org.apache.lucene.index.FieldInfo;
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.index.StoredFieldVisitor;
+import org.apache.lucene.util.BytesRef;
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.index.mapper.BlockLoader;
+import org.elasticsearch.index.mapper.IgnoredSourceFieldMapper;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+class IgnoredSourceFieldLoader extends StoredFieldLoader {
+    private final boolean forceSequentialReader;
+    private final Map<String, Set<String>> potentialFieldsInIgnoreSource;
+    private final Set<String> fieldNames;
+
+    IgnoredSourceFieldLoader(BlockLoader.FieldsSpec spec, boolean forceSequentialReader) {
+        fieldNames = new HashSet<>(spec.ignoredFieldsSpec().requiredIgnoredFields());
+        this.forceSequentialReader = forceSequentialReader;
+
+        HashMap<String, Set<String>> potentialFieldsInIgnoreSource = new HashMap<>();
+        for (String requiredIgnoredField : spec.ignoredFieldsSpec().requiredIgnoredFields()) {
+            for (String potentialStoredField : spec.ignoredFieldsSpec().format().requiredStoredFields(requiredIgnoredField)) {
+                potentialFieldsInIgnoreSource.computeIfAbsent(potentialStoredField, k -> new HashSet<>()).add(requiredIgnoredField);
+            }
+        }
```

Comment:
I hoping that we can optimize/simplify in a follow up change. This logic is now required for fields with dots and then we don't know which field part there is actually a stored field for. 

Ideally we would know based from the mapping what stored field we need. For example if  `attributes.host.ip` is requested and no `attributes` field is mapped, then the stored field should be `_ignored_source.attributes`. If attributes is mapped, but host isn't then the required stored field to load would be `_ignored_source.attributes.host`.

I also think `requiredStoredFields(...)` also always include the `_doc` field (via `FallbackSyntheticSourceBlockLoader.splitIntoFieldPaths(...)`)?

Would be great if we need less maps and no sets in `SFV`.